In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

# Elden Ring

Elden Ring, released February 24th 2022 by From Software, is an action-RPG in From Soft's "Souls" series which dates back to 2009's Demon's Souls. Elden Ring is an open world game with a heavy focus on combat, abstract storytelling that players can piece together. 

The game has Player versus Environment (PvE) combat in the open world, small dungeons which end in a boss encounter, and lengthy Legacy Dungeons which host multiple bosses and harken back to the levels of the previous Souls games (which are not open world). This is blended with a rich Player versus Player (PvP) combat which has been another staple of the Souls series from its inception. Players may invite other players into their world, be they friends or strangers who offer help to anyone the game may connect them with, to fight through the game's PvE content with proportionately stronger enemies. The main cost of this though, is this opens them up to players invading their world to engage in PvP. Worlds will have up to four players: one host, up to two co-opers, and one invader (or two if the host uses an item that draws in repeated invasions). Alternatively, players can summon others into their world for 1v1 duels, and these matches were typically held in the round promenade before the entrance to a wizard school legacy dungeon. However, December 7th's version 1.08 changed everything by adding three PvP colosseums which provide swift matchups for duels, as well as 2v2 and 3v3 combat and 2, 4, or 6 player free for all battles.  

Souls games host an abstract but rich story which players must infer from the details of the game like environmental storytelling, the uncommon piece of dialogue, item descriptions and so on. This has fostered a community that scrutinizes every aspect of these games, with YouTubers such as illusorywall and Limit Breakers delving deep into the game's mechanics (the former is where I found this spreadsheet), Zullie the Witch who data mines and uses cheat engine tools to garner additional story details and other interesting facts, and Sekiro Dubi who reconstructs cut content based on elements that are hidden in the game files. In addition the PvP community has a vested interest in knowing how weapons work mechanically, and the game's patch notes are about as vague as the story, so this community aspect persists through there.

In [2]:
MotData = pd.read_csv("ER - Motion Values and Attack Data (App Ver. 1.08) - Motion Values.csv")
PseData = pd.read_csv("ER - Motion Values and Attack Data (App Ver. 1.08) - Poise Damage - PvP.csv")

## Weapon Data

Elden Ring has 377 distinct weapons, and each weapon has a number of different attacks it can do based on different button inputs combined with context. The dataset's rows correspond to the weapons, and its columns correspond to each specific kind of attack. The dataset consists of over a dozen separate spreadsheets which have values that correspond to each attack. We will examine "Motion Values" and "Poise Damage - PvP". The former determines what percentage change to a weapon's base damage a particular move should have; generally slower moves are more powerful, moves later into combo chains deal more damage, and critical hits deal massive amounts of damage. The latter refers to how good attacks are at staggering other players in "Player versus Player" combat, with player poise being measured on a scale of 0-133.

These values are not necessarily listed in the game directly, rather they are the end result of some formula the game runs. In particular, patch 1.07 states "Improved poise damage in PvP for every weapon’s normal attack, except for Skills and long-ranged weapons.", so the poise damage players deal to each other is higher than what players deal to enemies controlled by the game. Dataminers have been compiling all of these numbers, and we have user Cryptid Tracker to thank for these data.

In [3]:
sum(MotData["Weapon"].notnull())

377

In [4]:
MotData

Weapon Class               Weapon  1h R1 1  1h R1 2  1h R1 3  1h R1 4  \
0            NaN                  NaN      NaN      NaN      NaN      NaN   
1         Dagger               Dagger    100.0    101.0    102.0    103.0   
2         Dagger      Parrying Dagger    100.0    101.0    102.0    103.0   
3         Dagger          Miséricorde    100.0    101.0    102.0    103.0   
4         Dagger          Great Knife    100.0    101.0    102.0    103.0   
..           ...                  ...      ...      ...      ...      ...   
406  Greatshield    Ant's Skull Plate    100.0    102.0    105.0      NaN   
407  Greatshield  Erdtree Greatshield    100.0    102.0    105.0      NaN   
408  Greatshield     Jellyfish Shield    100.0    102.0    105.0      NaN   
409  Greatshield        Visage Shield    100.0    102.0    105.0      NaN   
410  Greatshield      One-Eyed Shield    100.0    102.0    105.0      NaN   

     1h R1 5  1h R1 6 1h R2 1 1h R2 2  ... Mounted R1 3 Mounted R2  \
0        NaN      NaN     NaN     NaN  ...          NaN        NaN   
1      104.0    110.0     120     120  ...        110.0      135.0   
2      104.0    110.0     120     120  ...        110.0      135.0   
3      104.0    110.0     120     120  ...        110.0      135.0   
4      104.0    110.0     120     120  ...        110.0      135.0   
..       ...      ...     ...     ...  ...          ...        ...   
406      NaN      NaN     125     125  ...        110.0      125.0   
407      NaN      NaN     125     125  ...        110.0      125.0   
408      NaN      NaN     125     125  ...        110.0      125.0   
409      NaN      NaN     125     125  ...        110.0      125.0   
410      NaN      NaN     125     125  ...        110.0      125.0   

     Mounted R2 Charging Mounted Charged R2  Mounted L1 1  Mounted L1 2  \
0                    NaN                NaN           NaN           NaN   
1                   90.0                170         110.0         110.0   
2                   90.0                170         110.0         110.0   
3                   90.0                170         110.0         110.0   
4                   90.0                170         110.0         110.0   
..                   ...                ...           ...           ...   
406                100.0                165         110.0         110.0   
407                100.0                165         110.0         110.0   
408                100.0                165         110.0         110.0   
409                100.0                165         110.0         110.0   
410                100.0                165         110.0         110.0   

    Mounted L1 3 Mounted L2  Mounted L2 Charging Mounted Charged L2  
0            NaN        NaN                  NaN                NaN  
1          110.0      135.0                 90.0                170  
2          110.0      135.0                 90.0                170  
3          110.0      135.0                 90.0                170  
4          110.0      135.0                 90.0                170  
..           ...        ...                  ...                ...  
406        110.0      125.0                100.0                165  
407        110.0      125.0                100.0                165  
408        110.0      125.0                100.0                165  
409        110.0      125.0                100.0                165  
410        110.0      125.0                100.0                165  

[411 rows x 72 columns]

In [5]:
WeaponNames = MotData["Weapon"].to_numpy()

In [6]:
print(np.where(WeaponNames == "Godskin Peeler")[0][0])
print(np.where(WeaponNames == "Academy Glintstone Staff")[0][0])
print(np.where(WeaponNames == "Wing of Astel")[0][0])

122
306
98


In [7]:
MotData.size

29592

In [8]:
MotData.shape

(411, 72)

## Struggles with the Dataset

### NaN

The dataset has many null entries, which pandas interprets as Not a Number (NaN). These exist in empty rows placed to separate different weapon classes (Daggers, Straight Swords, Thrusting Swords, Sheilds, Glintstone Staves, etc.), so that the google sheet version is more readable. Second, many weapon classes cannot perform certain attacks. Whenever I work with the data, I must include some functionality to exclude those values.

### Multi Attacks

Some weapons in Elden Ring do multiple separate attacks with a single input, like holding a twinblade in two hands will have your player swing it around multiple times and each will count as a separate hit, any amount of which can hit your foes. Additionally some weapons have an additional projectile or Area of Effect (AoE) attack. The spreadsheet formats this as `50 + 50 + 50 + 165`, and `80 + (150 Projectile) + 105 + (150 Projectile)`. Clearly, these cannot be parsed as integers by the data analysis tools I want to utilize. I decided to simply sum all numbers contained in these strings whenever I encountered them, and I had to make code specifically to do this.

In [9]:
with pd.option_context('display.max_rows', None,
                       'display.max_columns', None,
                       'display.precision', 3,
                       ):
    print(MotData.iloc[122])

Weapon Class                       Twinblade
Weapon                        Godskin Peeler
1h R1 1                                100.0
1h R1 2                                101.0
1h R1 3                                102.0
1h R1 4                                103.0
1h R1 5                                110.0
1h R1 6                                  NaN
1h R2 1                                  125
1h R2 2                                  125
1h Charged R2 1                          160
1h Charged R2 2                          160
1h Running R1                          105.0
1h Running R2                       50 + 115
1h Rolling R1                           90.0
1h Backstep R1                          90.0
1h Jumping R1                            107
1h Jumping R2                            127
1h Guard Counter                       125.0
2h R1 1                             50 + 103
2h R1 2                             50 + 104
2h R1 3                                  115
2h R1 4   

In [10]:
with pd.option_context('display.max_rows', None,
                       'display.max_columns', None,
                       'display.precision', 3,
                       ):
    print(MotData.iloc[306])

Weapon Class                     Glintstone Staff
Weapon                   Academy Glintstone Staff
1h R1 1                                     100.0
1h R1 2                                     101.0
1h R1 3                                     102.0
1h R1 4                                       NaN
1h R1 5                                       NaN
1h R1 6                                       NaN
1h R2 1                                       NaN
1h R2 2                                       NaN
1h Charged R2 1                               NaN
1h Charged R2 2                               NaN
1h Running R1                                 NaN
1h Running R2                                 NaN
1h Rolling R1                                 NaN
1h Backstep R1                                NaN
1h Jumping R1                                 107
1h Jumping R2                                 117
1h Guard Counter                              NaN
2h R1 1                                       105


In [11]:
with pd.option_context('display.max_rows', None,
                       'display.max_columns', None,
                       'display.precision', 3,
                       ):
    print(MotData.iloc[98])

Weapon Class                                               Curved Sword
Weapon                                                    Wing of Astel
1h R1 1                                                           100.0
1h R1 2                                                           101.0
1h R1 3                                                           102.0
1h R1 4                                                           103.0
1h R1 5                                                           104.0
1h R1 6                                                           110.0
1h R2 1                                           70 + (150 Projectile)
1h R2 2                                           70 + (150 Projectile)
1h Charged R2 1           70 + (150 Projectile) + 80 + (150 Projectile)
1h Charged R2 2           70 + (150 Projectile) + 80 + (150 Projectile)
1h Running R1                                                     105.0
1h Running R2                                                   

In [12]:
def stringconv(input):
    if(type(input) != str):
        return float(input)
    if(input.replace(".", "", 1).isdigit()):
        return float(input)
    sum = 0.0
    tempstr = ""
    for x in input:
        if x.isdigit() or x == '.':
            tempstr+=x
            continue
        if tempstr == "":
            continue
        sum += float(tempstr)
        tempstr = ""
    if tempstr == "":
        return sum
    sum += float(tempstr)
    tempstr = ""
    return sum

In [13]:
stringconv("103294.98")

103294.98

In [14]:
stringconv("115 + (57.5 Projectile) + 230 + (57.5 Projectile)")

460.0

In [15]:
MotData["1h R2 1"] = MotData["1h R2 1"].apply(stringconv)
PseData["1h R2 1"] = PseData["1h R2 1"].apply(stringconv)
MotData["1h Running R1"] = MotData["1h Running R1"].apply(stringconv)
PseData["1h Running R1"] = PseData["1h Running R1"].apply(stringconv)
MotData["1h Running R2"] = MotData["1h Running R2"].apply(stringconv)
PseData["1h Running R2"] = PseData["1h Running R2"].apply(stringconv)

## Poise Breakpoints

### Poise

Poise is a value from 0 to 133 which determines how hard it is for enemy attacks to stagger your player character, breifly preventing you from taking more actions and giving a small window for unimpeded enemy attacks. Each attack deals a certain amount of "poise damage", and when your character has been dealt enough poise damage, they will stagger. If the player has not been dealt poise damage for 60 seconds or gets staggered, the stored amount of poise damage will reset to zero.

Your character's armor, separated into headgear, chest, gauntlets, and leggings, will contribute 0-100 poise in total. The highest poise armor sets are the "Bull-Goat Set" giving 100 poise, the "Fire Prelate Set" at 96 poise, the "Omen Set" at 91 poise, "Lionel's Set" at 86 poise, and the "Veteran's Set" at 80 poise. In addition there is the "Bull-Goat's Talisman" which reduces incoming poise damage by 25%, which the game displays by saying you have 33% more poise, bringing the maximum to 133.

There are temporary effects which increase poise, but as you will see we don't need to account for them in the following analysis. First, Weapon Skill "Endure" will give you infinite poise for 3.5 seconds. The Flask of Woundrous Physick can hold two crystal tears, and can only be used once per fight, so "Leaden Hardtear" gives infinite poise for 10 seconds. The consumable "Ironjar Aromatic" greatly slows the player's speed and makes attacks bounce off of them, at the cost of slowing the player to a crawl. Of little import is (Radiant) Baldachin's Blessing increases poise by 50% (81%) for 15 (70) seconds, these items being far too tedious to obtain repeatedly to see use.

### Limitations of higher poise

Why wouldn't all players simply have 133 poise? The answer is multifaceted. Higher poise armor will be heavier, which has a complex statistical interplay, players may choose armor for fashion rather than poise, and players have to find sets of armor out in the world. 

#### Equip Load
Players have the statistic "Equip Load" which determines how much the sum total of held weapons, armor, and talismans will affect them. Being below 30%, 70%, and 100% or above 100% equip load constitute four tiers of mobility, with lighter tiers granting faster actions and especially faster dodge rolls, while the heaviest tier restricts player to a slow walk and disables dodge rolling. Equip Load scales with the player statistic "Endurance" which also governs stamina i.e. how many actions a player can do at once. Increasing Endurance comes at the cost of increasing statistics which make the player deal more damage, Vigor which governs player health, and Mind which governs mana i.e. how many spells and weapon arts a player can cast. Thus there is a tradition of "tank" builds wearing the heaviest armor and using slower but high damage weapons, while sorcerers tend to wear lighter armor and invest into other stats so their spells can deal absurd amounts of damage. Most players aim to have just below 70% equip load to be in the medium tier and have a decent amount of armor. 

#### Fashion
Secondarily, fashion is another major factor in what armor players use. The Bull-Goat set and the Omen set are considered ugly by many, turning them off from using it. The Fire Prelate chestpiece and helmet also have a very specific look that not all players would want, but the leggings and gauntlets can blend nicely into most attires. Alternatively, players may just have a fondness for how particular pieces of armor look, or they may be "cosplaying" as a particular enemy in the game by using their armor or as a particular character from outside media.

#### Non-poise benefits
Poise is far from the only statistic of armor. Each piece of armor has varying resistances for different types of damage, and increases the player's resistances to certain status effects (bleed, scarlet rot, madness, etc.) in varying amounts. In addition, many helmets grant other benefits like boosting the aforementioned player statistics, and some sets of armor make thematically associated spells stronger (which mechanically encourages cosplaying as certain enemies).

#### Obtaining armor
The higher poise armors are the hardest to obtain, two of the five I listed require you to get into the late game area of the Capital, and a further two require you to complete that area. This leaves the Bull-Goat set as the one you have earliest access in theory, but it requires your character to join a murder cult and assassinate the NPC "Great Horned Targoth" who wears the armor. Furthermore, this quest can be locked out by either killing the Demigod leader of the cult, upon which it dissolves, or choosing to kill Patches when he appears as a boss early in the game, who asks you to spare him at half health. The Lionel's Set is in a very obvious location, so most players will obtain it, and the Fire Prelate Set is a random drop from a strong enemy that players must go past and can kill as many times as they like until they get the complete set.

#### Talismans
Players can use up to four talismans at once, and one of these can be the Bull-Goat's Talisman. The game has many talismans useful to all players, and many that benefit specific builds, so the choice of whether or not to use that specific talisman is significant, and among talismans it has a low equip load cost. Players may opt into using it if they are using lighter armor for the aforementioned reasons, or opt out of using it because they prefer other talismans. The item is relatively easy to obtain.


### Breakpoints
All of this exposition is to say, players want to know what values of poise they can achieve which would allow them to withstand a wide variety of attacks. The community refers to "Poise Breakpoints" as levels of poise which will protect players from being staggered by a certain class of attack. 

#### PvE breakpoints
Enemies all deal either 50 or 120 poise damage with every attack. In addition there are hostile NPCs that will deal the same highly variable levels of poise damage damage we will discuss next. Thus for most "Player versus Environment" combat, it really only matters whether players at 51 poise or above, at 101 poise or above (to sustain two 50 poise attacks) or at 121-133 poise (only the three strongest sets of armor with Bull-Goat's Talisman grant this much poise). 

#### PvP breakpoints


We want to guess what the best poise values a player can have in PvP to be able to withstand a high proportion of attacks without being staggered.

In [16]:
PseArray = np.concatenate((np.array(PseData["1h R1 1"].dropna()), np.array(PseData["1h R2 1"].dropna())))

In [17]:
np.percentile(PseArray, 2)

40.5